### PR TITLE
BioSQL doesn't need to subclass Seq

### PR DESCRIPTION
**Pull request by mdehoon**
- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit`` locally,
and understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

With the recent changes in `Bio.Seq`, it is no longer necessary for BioSQL to define its own `DBSeq` class as a subclass of the `Seq` class in `Bio.Seq`. A data provider class suffices. This simplifies the code in `BioSQL`, and avoids the `Seq` class in `Bio.Seq` to be aware of the implementation details of its subclass.

### CODE REVIEW

File: BioSQL/BioSeq.py [hunk 1]
```diff
@@ -23,17 +23,17 @@
...
-        """Create a new DBSeq object referring to a BioSQL entry.
+    def __init__(self, primary_id, adaptor, start=0, length=0):
+        """Create a new _BioSQLSequenceData object referring to a BioSQL entry.
 
-        You wouldn't normally create a DBSeq object yourself, this is done
-        for you when retrieving a DBSeqRecord object from the database.
+        You wouldn't normally create a _BioSQLSequenceData object yourself,
+        this is done for you when retrieving a DBSeqRecord object from the
+        database, which creates a Seq object using a _BioSQLSequenceData
+        instance as the data provider..
```
Comment (peterjc) [inline BioSQL/BioSeq.py:35]: Extra dot

File: BioSQL/BioSeq.py [hunk 2]
```diff
@@ -43,147 +43,45 @@ def __len__(self):
...
+            return full[:: step].encode("ASCII")
 
     def __bytes__(self):
         """Return the full sequence as bytes."""
-        return str(self).encode("ASCII")
-
-    def __str__(self):
-        """Return the full sequence as a python string."""
-        return self.adaptor.get_subseq_as_string(
+        sequence = self.adaptor.get_subseq_as_string(
```
Comment (peterjc) [inline BioSQL/BioSeq.py:81]: Is this bytes or unicode?

File: BioSQL/BioSeqDatabase.py [hunk 3]
```diff
@@ -691,7 +691,7 @@ def get_Seq_by_id(self, name):
         Example: seq_rec = db.get_Seq_by_id('ROA1_HUMAN')
 
         The name of this method is misleading since it returns a DBSeqRecord
-        rather than a DBSeq ojbect, and presumably was to mirror BioPerl.
+        rather than a Seq ojbect, and presumably was to mirror BioPerl.
```
Comment (peterjc) [inline BioSQL/BioSeqDatabase.py:694]: Old typo ojbect -> object (and the following cases too)

**Comment by peterjc** [comment]
No one should be creating their own instances of the DBSeq class (as per the documentation), so this should be OK in principle.

See minor comments.

**Comment by peterjc** [comment]
```
$ black --check --diff .
--- BioSQL/BioSeq.py	2020-12-04 18:34:30.413205 +0000
+++ BioSQL/BioSeq.py	2020-12-04 18:35:40.343201 +0000
@@ -72,11 +72,11 @@
         else:
             # Will have to extract the sequence because of the stride
             full = self.adaptor.get_subseq_as_string(
                 self.primary_id, self.start + start, self.start + end
             )
-            return full[:: step].encode("ASCII")
+            return full[::step].encode("ASCII")
 
     def __bytes__(self):
         """Return the full sequence as bytes."""
         sequence = self.adaptor.get_subseq_as_string(
             self.primary_id, self.start, self.start + self._length
would reformat BioSQL/BioSeq.py
Oh no! 💥 💔 💥
1 file would be reformatted, 563 files would be left unchanged.
```

File: BioSQL/BioSeq.py (hunk 2 above)
Comment (mdehoon) [reply BioSQL/BioSeq.py:81]: bytes

Comment (peterjc) [reply BioSQL/BioSeq.py:81]: This method ``get_subseq_as_string`` returns the value from the SQL query "SELECT SUBSTR(seq, %s, %s) FROM biosequence WHERE bioentry_id = %s" but it is not clear if this is going to be unicode or bytes.

Prior to your pull request, this was used for ``__str__`` suggesting it was a unicode? Now you say is bytes. Does that mean you've found and fixed another Python 2/3 oversight?

Or is this more complicated in that bytes vs unicode depends on the SQL back end and/or adaptor (e.g. one of the MySQL interfaces, SQLite3, etc)?

Comment (mdehoon) [reply BioSQL/BioSeq.py:81]: `sequence` is `str`, which is then transformed to `bytes` on the next line by calling `encode`

Comment (peterjc) [reply BioSQL/BioSeq.py:81]: Clear now, thanks. That final encode wasn't visually close in the diff view.

**Review by peterjc — APPROVED** [review]

**Pull request merged**